src/main/kotlin/org/rust/ide/annotator/RsSyntaxErrorsAnnotator.kt

```diff
@@ -217,6 +252,40 @@ private fun checkTypeParameterList(holder: AnnotationHolder, element: RsTypePara
     }
 }
 
+private fun checkTypeArgumentList(holder: AnnotationHolder, args: RsTypeArgumentList) {
+    var kind = TypeArgumentKind.LIFETIME
+    args.forEachChild { child ->
+        val newKind = TypeArgumentKind.forType(child) ?: return@forEachChild
+        if (newKind.canStandAfter(kind)) {
+            kind = newKind
+        } else {
+            val newStateName = newKind.argumentNameCapitalized
+            holder.createErrorAnnotation(child, "$newStateName must be declared prior to ${kind.argumentName}")
+        }
+    }
+}
+
+private enum class TypeArgumentKind(private val elementClass: KClass<*>, val argumentName: String) {
+    LIFETIME(RsLifetime::class, "lifetime arguments"),
+    TYPE(RsTypeReference::class, "type arguments"),
+    CONST(RsExpr::class, "const arguments"),
+    ASSOC(RsAssocTypeBinding::class, "associated type bindings");
+
+    val argumentNameCapitalized: String
+        get() = StringUtil.capitalize(argumentName)
+
+    fun canStandAfter(prevArgument: TypeArgumentKind): Boolean =
+        ordinal >= prevArgument.ordinal
+
+    data class Result(val newState: TypeArgumentKind, val allowed: Boolean)
+
+    companion object {
+        private val VALUES = values()
```

Comment:
Why do we need this constant?

Reply:
`values()` allocates a copy of the array on each invocation...